[Valgrind] Fix some jumps depending on uninitialized values

neo1973 opened the pull request:
## Description
See title and Valgrind output in commit messages.

## Motivation and context
Remove undefined behaviour and maybe improve stability.

## How has this been tested?
Errors in Valgrind are gone, Kodi still runs :)

## What is the effect on users?
None.

## Screenshots (if appropriate):
None

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

neo1973 committed:
XBDateTime: Fix jump on uninitialized value

Valgrind:

==17974== Conditional jump or move depends on uninitialised value(s)
==17974==    at 0x5962156: isdst_differ (mktime.c:191)
==17974==    by 0x5962156: __mktime_internal (mktime.c:426)
==17974==    by 0x1A617E6: CDateTime::GetAsTm(tm&) const (xbmc/xbmc/XBDateTime.cpp:829)
==17974==    by 0xEE1F9D: TestDateTime_TmOperators_Test::TestBody() (xbmc/xbmc/test/TestDateTime.cpp:85)
==17974==    by 0x48CFB3B: UnknownInlinedFun (gtest.cc:2621)
==17974==    by 0x48CFB3B: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [clone .constprop.0] (gtest.cc:2657)
==17974==    by 0x48BC197: UnknownInlinedFun (gtest.cc:2696)
==17974==    by 0x48BC197: testing::Test::Run() (gtest.cc:2686)
==17974==    by 0x48BC3A6: testing::TestInfo::Run() (gtest.cc:2845)
==17974==    by 0x48BC503: UnknownInlinedFun (gtest.cc:3004)
==17974==    by 0x48BC503: testing::TestSuite::Run() (gtest.cc:2977)
==17974==    by 0x48C8A6E: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:5890)
==17974==    by 0x48C76A8: UnknownInlinedFun (gtest.cc:2621)
==17974==    by 0x48C76A8: UnknownInlinedFun (gtest.cc:2657)
==17974==    by 0x48C76A8: testing::UnitTest::Run() (gtest.cc:5455)
==17974==    by 0xDA0A50: RUN_ALL_TESTS (gtest.h:2314)
==17974==    by 0xDA0A50: main (xbmc/xbmc/test/xbmc-test.cpp:29)

neo1973 committed:
VideoPlayer: Fix jump on uninitialized value

Valgrind example:

==25638== Conditional jump or move depends on uninitialised value(s)
==25638==    at 0x155616D: CVideoPlayer::GetVideoStreamInfo(int, VideoStreamInfo&) const (xbmc/xbmc/cores/VideoPlayer/VideoPlayer.cpp:5203)
==25638==    by 0x1AA8AC4: CApplicationPlayer::GetVideoStreamInfo(int, VideoStreamInfo&) const (xbmc/xbmc/application/ApplicationPlayer.cpp:586)
==25638==    by 0x1C65CD6: CGUIInfoManager::UpdateAVInfo() (xbmc/xbmc/GUIInfoManager.cpp:11064)
==25638==    by 0x19E41BF: CGUIWindowManager::FrameMove() (xbmc/xbmc/guilib/GUIWindowManager.cpp:1377)
==25638==    by 0x1A9652C: CApplication::FrameMove(bool, bool) (xbmc/xbmc/application/Application.cpp:1868)
==25638==    by 0x1A96BA4: CApplication::Run() (xbmc/xbmc/application/Application.cpp:1912)
==25638==    by 0x188B1DC: XBMC_Run (xbmc/xbmc/platform/xbmc.cpp:61)
==25638==    by 0x1210243: main (xbmc/xbmc/platform/posix/main.cpp:71)
==25638==  Uninitialised value was created by a heap allocation
==25638==    at 0x4843003: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==25638==    by 0x22DDEBC: allocate (new_allocator.h:137)
==25638==    by 0x22DDEBC: allocate (alloc_traits.h:464)
==25638==    by 0x22DDEBC: __allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<CVideoPlayer, std::allocator<void>, (__gnu_cxx::_Lock_policy)2> > > (allocated_ptr.h:98)
==25638==    by 0x22DDEBC: __shared_count<CVideoPlayer, std::allocator<void>, IPlayerCallback &> (shared_ptr_base.h:969)
==25638==    by 0x22DDEBC: __shared_ptr<std::allocator<void>, IPlayerCallback &> (shared_ptr_base.h:1712)
==25638==    by 0x22DDEBC: shared_ptr<std::allocator<void>, IPlayerCallback &> (shared_ptr.h:464)
==25638==    by 0x22DDEBC: make_shared<CVideoPlayer, IPlayerCallback &> (shared_ptr.h:1009)
==25638==    by 0x22DDEBC: CPlayerCoreConfig::CreatePlayer(IPlayerCallback&) const (xbmc/xbmc/cores/playercorefactory/PlayerCoreConfig.cpp:49)
==25638==    by 0x22E0413: CPlayerCoreFactory::CreatePlayer(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IPlayerCallback&) const (xbmc/xbmc/cores/playercorefactory/PlayerCoreFactory.cpp:64)
==25638==    by 0x1AA3E13: CApplicationPlayer::CreatePlayer(CPlayerCoreFactory const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IPlayerCallback&) (xbmc/xbmc/application/ApplicationPlayer.cpp:70)
==25638==    by 0x1AA45D5: CApplicationPlayer::OpenFile(CFileItem const&, CPlayerOptions const&, CPlayerCoreFactory const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, IPlayerCallback&) (xbmc/xbmc/application/ApplicationPlayer.cpp:139)
==25638==    by 0x1A925FF: CApplication::PlayFile(CFileItem, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) (xbmc/xbmc/application/Application.cpp:2545)
==25638==    by 0x1C8AFCE: PLAYLIST::CPlayListPlayer::Play(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool) (xbmc/xbmc/PlayListPlayer.cpp:337)
==25638==    by 0x1C8BC2E: PLAYLIST::CPlayListPlayer::Play(std::shared_ptr<CFileItem> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/PlayListPlayer.cpp:299)
==25638==    by 0x158C2FA: CGUIWindowVideoBase::PlayMovie(CFileItem const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/video/windows/GUIWindowVideoBase.cpp:1037)
==25638==    by 0x158C10E: CGUIWindowVideoBase::OnPlayMedia(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/video/windows/GUIWindowVideoBase.cpp:1010)
==25638==    by 0x1707EB4: CGUIMediaWindow::OnClick(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/windows/GUIMediaWindow.cpp:1179)
==25638==    by 0x15A0293: CGUIWindowVideoNav::OnClick(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/video/windows/GUIWindowVideoNav.cpp:1169)

neo1973 committed:
VideoPlayerCodec: Fix jump on uninitialized value

Valgrind:

==28230== Conditional jump or move depends on uninitialised value(s)
==28230==    at 0x22F0007: VideoPlayerCodec::ReadPCM(unsigned char*, unsigned long, unsigned long*) (xbmc/xbmc/cores/paplayer/VideoPlayerCodec.cpp:358)
==28230==    by 0x22EF540: VideoPlayerCodec::Init(CFileItem const&, unsigned int) (xbmc/xbmc/cores/paplayer/VideoPlayerCodec.cpp:190)
==28230==    by 0x22E68A3: CAudioDecoder::Create(CFileItem const&, long) (xbmc/xbmc/cores/paplayer/AudioDecoder.cpp:84)
==28230==    by 0x22EAE16: PAPlayer::QueueNextFileEx(CFileItem const&, bool) (xbmc/xbmc/cores/paplayer/PAPlayer.cpp:336)
==28230==    by 0x22ECEB2: operator() (xbmc/xbmc/cores/paplayer/PAPlayer.cpp:228)
==28230==    by 0x22ECEB2: CLambdaJob<PAPlayer::OpenFile(CFileItem const&, CPlayerOptions const&)::$_0>::DoWork() (xbmc/xbmc/utils/JobManager.h:39)
==28230==    by 0x1793278: CJobWorker::Process() (xbmc/xbmc/utils/JobManager.cpp:55)
==28230==    by 0x1844B05: CThread::Action() (xbmc/xbmc/threads/Thread.cpp:267)
==28230==    by 0x1844E6C: operator() (xbmc/xbmc/threads/Thread.cpp:138)
==28230==    by 0x1844E6C: __invoke_impl<void, (lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:61)
==28230==    by 0x1844E6C: __invoke<(lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:96)
==28230==    by 0x1844E6C: _M_invoke<0UL, 1UL, 2UL> (std_thread.h:258)
==28230==    by 0x1844E6C: operator() (std_thread.h:265)
==28230==    by 0x1844E6C: std::thread::_State_impl<std::thread::_Invoker<std::tuple<CThread::Create(bool)::$_0, CThread*, std::promise<bool> > > >::_M_run() (std_thread.h:210)
==28230==    by 0x6F71282: execute_native_thread_routine (thread.cc:82)
==28230==    by 0x58B98FC: start_thread (pthread_create.c:442)
==28230==    by 0x593AF33: clone (clone.S:100)
==28230==  Uninitialised value was created by a heap allocation
==28230==    at 0x4843003: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==28230==    by 0x22E806B: CodecFactory::CreateCodec(CURL const&) (xbmc/xbmc/cores/paplayer/CodecFactory.cpp:41)
==28230==    by 0x22E88B3: CodecFactory::CreateCodecDemux(CFileItem const&, unsigned int) (xbmc/xbmc/cores/paplayer/CodecFactory.cpp:110)
==28230==    by 0x22E6885: CAudioDecoder::Create(CFileItem const&, long) (xbmc/xbmc/cores/paplayer/AudioDecoder.cpp:82)
==28230==    by 0x22EAE16: PAPlayer::QueueNextFileEx(CFileItem const&, bool) (xbmc/xbmc/cores/paplayer/PAPlayer.cpp:336)
==28230==    by 0x22ECEB2: operator() (xbmc/xbmc/cores/paplayer/PAPlayer.cpp:228)
==28230==    by 0x22ECEB2: CLambdaJob<PAPlayer::OpenFile(CFileItem const&, CPlayerOptions const&)::$_0>::DoWork() (xbmc/xbmc/utils/JobManager.h:39)
==28230==    by 0x1793278: CJobWorker::Process() (xbmc/xbmc/utils/JobManager.cpp:55)
==28230==    by 0x1844B05: CThread::Action() (xbmc/xbmc/threads/Thread.cpp:267)
==28230==    by 0x1844E6C: operator() (xbmc/xbmc/threads/Thread.cpp:138)
==28230==    by 0x1844E6C: __invoke_impl<void, (lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:61)
==28230==    by 0x1844E6C: __invoke<(lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:96)
==28230==    by 0x1844E6C: _M_invoke<0UL, 1UL, 2UL> (std_thread.h:258)
==28230==    by 0x1844E6C: operator() (std_thread.h:265)
==28230==    by 0x1844E6C: std::thread::_State_impl<std::thread::_Invoker<std::tuple<CThread::Create(bool)::$_0, CThread*, std::promise<bool> > > >::_M_run() (std_thread.h:210)
==28230==    by 0x6F71282: execute_native_thread_routine (thread.cc:82)
==28230==    by 0x58B98FC: start_thread (pthread_create.c:442)
==28230==    by 0x593AF33: clone (clone.S:100)

garbear commented on xbmc/cores/VideoPlayer/VideoPlayer.h:
```diff
@@ -488,10 +488,9 @@ class CVideoPlayer : public IPlayer, public CThread, public IVideoPlayer,
-    int m_subtitleIndex;
+    int m_videoIndex = -1;
+    int m_audioIndex = -1;
+    int m_subtitleIndex = -1;
```
We switched our code style a while ago, can you brace-initialize member variables? Like:

```c++
int m_videoIndex{-1};
int m_audioIndex{-1};
int m_subtitleIndex{-1};
```

neo1973 replied:
No problem :relaxed:

garbear commented:
I'm glad people run valgrind, otherwise these would never get caught!

garbear commented on xbmc/cores/paplayer/VideoPlayerCodec.cpp:
```diff
@@ -29,7 +29,10 @@ VideoPlayerCodec::VideoPlayerCodec()
+  m_bCanSeek = false;
   m_pResampler = NULL;
+  m_audioFrame = {};
+  m_planes = 0;
```
Another style nit, we prefer to initialize member variables in the header file with the same brace initialization, because it works for all constructors.

garbear commented:
@lrusak Can you check the XBDateTime change?

howie-f commented on xbmc/cores/paplayer/VideoPlayerCodec.h:
```diff
@@ -41,24 +41,24 @@ class VideoPlayerCodec : public ICodec
+  ActiveAE::IAEResample* m_pResampler{};
+  DVDAudioFrame m_audioFrame{};
+  int m_planes{};
+  bool m_needConvert{};
```
next style nit/hint, i‘d prefer to be explicit with all those e.g. `{false}`

neo1973 replied:
I would agree but it looks like the code guidelines prefer the empty braces (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#54-initialization)? But then in a later example it's explicit again (https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#114-default-member-initialization)? :thinking:

howie-f replied:
uh yeah, please decide at your leisure :)

CastagnaIT replied:
imo i suggest to improve our docs to use only explicits init values
are more readable when you investigate for something

garbear replied:
I agree, I think that explicit values are better for when it's a primitive type, empty elsewhere, and avoided when possible, e.g. don't do `std::string example{}` because std::string has a default constructor. Doc improvement would be good.

thexai commented on xbmc/cores/paplayer/VideoPlayerCodec.h:
```diff
@@ -41,24 +41,24 @@ class VideoPlayerCodec : public ICodec
   CAEStreamInfo::DataType GetPassthroughStreamType(AVCodecID codecId, int samplerate, int profile);
 
-  CDVDDemux* m_pDemuxer;
+  CDVDDemux* m_pDemuxer{};
```
For raw pointers seems more clear use `{nullptr}`

Also is not the case here but consider this code:

```c++
    int num = 3;
    int* p_num = nullptr;

    p_num = &num;
```
then 

`p_num = {};`

this is same that `p_num = nullptr` and `num` still == 3

but 

`*p_num = {};`  

means `num == 0`  and `p_num` still holds address of `num`

This confusion has happened here: https://github.com/xbmc/xbmc/pull/21679/commits/2172f41d12c67df5f1887a722eac09ade29ac52c

garbear replied:
Well put. This suggestion is making it into our code guidelines at https://github.com/xbmc/xbmc/pull/22639

neo1973 commented:
Thanks for the feedback so far :relaxed: I found another issue, will try to fix it tomorrow.

neo1973 committed:
ZipFile: Fully initialize stat64 struct in Stat(...)

Example Valgrind output:

==21036== Conditional jump or move depends on uninitialised value(s)
==21036==    at 0x1CA484C: CTextureCacheJob::GetImageHash(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/TextureCacheJob.cpp:253)
==21036==    by 0x1CA3440: CTextureCacheJob::CacheTexture(std::unique_ptr<CTexture, std::default_delete<CTexture> >*) (xbmc/xbmc/TextureCacheJob.cpp:84)
==21036==    by 0x1CA0CB5: CTextureCache::CacheImage(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::unique_ptr<CTexture, std::default_delete<CTexture> >*, CTextureDetails*) (xbmc/xbmc/TextureCache.cpp:160)
==21036==    by 0x1C6AA9B: CImageLoader::DoWork() (xbmc/xbmc/GUILargeTextureManager.cpp:79)
==21036==    by 0x17945A8: CJobWorker::Process() (xbmc/xbmc/utils/JobManager.cpp:55)
==21036==    by 0x1845E35: CThread::Action() (xbmc/xbmc/threads/Thread.cpp:267)
==21036==    by 0x184619C: operator() (xbmc/xbmc/threads/Thread.cpp:138)
==21036==    by 0x184619C: __invoke_impl<void, (lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:61)
==21036==    by 0x184619C: __invoke<(lambda at /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/threads/Thread.cpp:94:32), CThread *, std::promise<bool> > (invoke.h:96)
==21036==    by 0x184619C: _M_invoke<0UL, 1UL, 2UL> (std_thread.h:258)
==21036==    by 0x184619C: operator() (std_thread.h:265)
==21036==    by 0x184619C: std::thread::_State_impl<std::thread::_Invoker<std::tuple<CThread::Create(bool)::$_0, CThread*, std::promise<bool> > > >::_M_run() (std_thread.h:210)
==21036==    by 0x6F71282: execute_native_thread_routine (thread.cc:82)
==21036==    by 0x58B98FC: start_thread (pthread_create.c:442)
==21036==    by 0x593AF33: clone (clone.S:100)
==21036==  Uninitialised value was created by a stack allocation
==21036==    at 0x1CA4659: CTextureCacheJob::GetImageHash(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/TextureCacheJob.cpp:242)

neo1973 committed:
VFSEntry: Fully initialize stat64 struct in Stat(...)

neo1973 committed:
PosixFile: Fully initialize stat64 struct in Stat(...)

neo1973 committed:
Win32File: Fully initialize stat64 struct in Stat(...)

neo1973 committed:
WinLibraryFile: Fully initialize stat64 struct in Stat(...)

neo1973 committed:
SlideShowPicture: Fix jump on uninitialized value

Example Valgrind output:

==20899== Conditional jump or move depends on uninitialised value(s)
==20899==    at 0x18B307D: CSlideShowPic::UpdateVertices(float*, float*, float const*, float const*, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/pictures/SlideShowPicture.cpp:283)
==20899==    by 0x18B42D9: CSlideShowPic::Process(unsigned int, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/pictures/SlideShowPicture.cpp:556)
==20899==    by 0x189834D: CGUIWindowSlideShow::Process(unsigned int, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/pictures/GUIWindowSlideShow.cpp:536)
==20899==    by 0x1965EAB: CGUIControl::DoProcess(unsigned int, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/guilib/GUIControl.cpp:143)
==20899==    by 0x19DA9FD: CGUIWindow::DoProcess(unsigned int, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/guilib/GUIWindow.cpp:330)
==20899==    by 0x197EF10: CGUIDialog::DoProcess(unsigned int, std::vector<CDirtyRegion, std::allocator<CDirtyRegion> >&) (xbmc/xbmc/guilib/GUIDialog.cpp:120)
==20899==    by 0x19E45C1: CGUIWindowManager::Process(unsigned int) (xbmc/xbmc/guilib/GUIWindowManager.cpp:1213)
==20899==    by 0x1A97897: CApplication::FrameMove(bool, bool) (xbmc/xbmc/application/Application.cpp:1866)
==20899==    by 0x19E572D: CGUIWindowManager::ProcessRenderLoop(bool) (xbmc/xbmc/guilib/GUIWindowManager.cpp:1411)
==20899==    by 0x197F20F: CGUIDialog::Open_Internal(bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/guilib/GUIDialog.cpp:179)
==20899==    by 0x197F307: CGUIDialog::Open(bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (xbmc/xbmc/guilib/GUIDialog.cpp:201)
==20899==    by 0x19E2E83: CGUIWindowManager::ActivateWindow_Internal(int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, bool, bool) (xbmc/xbmc/guilib/GUIWindowManager.cpp:848)
==20899==  Uninitialised value was created by a heap allocation
==20899==    at 0x4843003: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==20899==    by 0x19DF8A7: CGUIWindowManager::CreateWindows() (xbmc/xbmc/guilib/GUIWindowManager.cpp:298)
==20899==    by 0x1A8ECBA: CApplication::Initialize() (xbmc/xbmc/application/Application.cpp:708)
==20899==    by 0x188C514: XBMC_Run (xbmc/xbmc/platform/xbmc.cpp:43)
==20899==    by 0x12114F3: main (xbmc/xbmc/platform/posix/main.cpp:71)

garbear reviewed (approved):
Everything looks good from my end at this point. Good luck hunting that last remaining issue.

howie-f reviewed (approved):
nice and well documented in the commit messages. :+1:

garbear commented:
> nice and well documented in the commit messages. :+1: 

I didn't notice that, good job indeed.

fuzzard commented on xbmc/XBDateTime.cpp:
```diff
@@ -818,13 +818,15 @@ void CDateTime::GetAsTm(tm& time) const
   time.tm_hour = st.hour;
   time.tm_min = st.minute;
   time.tm_sec = st.second;
+  time.tm_isdst = -1;
```
My only query about this PR is this change.

https://github.com/xbmc/xbmc/pull/21660

The blanket -1 was said to not really be accurate, and may be better off with some time zone bias applied.

neo1973 replied:
I'm 99 percent certain that this change is correct :smiley: 

First, lets see what's wrong with the original code. To observe the issue you can set you PCs time zone to something in the southern hemisphere (I tried Auckland). If you then run the unit tests you will see these tests failing:
```
[  FAILED  ] TestDateTime.TmOperators
[  FAILED  ] TestDateTime.GetAsTm
```
What goes wrong? Let's look at the `GetAsTm` test:
https://github.com/xbmc/xbmc/blob/9104aa0af563f06dd6e4ed763b90f570ab556941/xbmc/test/TestDateTime.cpp#L543-L552
Here `CDateTime` is set to a time in May which doesn't observe daylight saving time in the southern hemisphere. When now calling `CDateTime::getAsTm` due to the missing initialization the `tm::tm_isdst` is likely some value > 0. Then `mktime` is called to normalize the `struct tm`. Because `tm::tm_isdst > 0` but no daylight saving active at the specified date, `mktime` subtracts an hour and sets `tm_isdst` to 0. But the `KODI::TIME::SystemTime` already was accounting for non daylight saving and we are now one hour off.

By initializing `tm_isdst` to `-1` we tell ``mktime to not adjust the time and to only set `tm_isdst` to the correct value.

I'll extend the unit tests so that this is caught independently of the local timezone in which the tests are run.

basilgello replied:
I'd postpone this pretty minor fix until we get cdatetime topic in. I have next week full with paid work but after that let me fix the freebsd failing test.

basilgello replied:
@neo1973 can you please have a look at cdatetime topic (without rebase) and check if anything from this PR is applicable there?

neo1973 replied:
> I'd postpone this pretty minor fix until we get cdatetime topic in. I have next week full with paid work but after that let me fix the freebsd failing test.

While the CDateTime changes will make this commit obsolete I still think it doesn't hurt do merge it now. Also it would make for a cleaner backport if this first goes to master.

> can you please have a look at cdatetime topic (without rebase) and check if anything from this PR is applicable there?

The only overlap is the `CDateTime::GetAsTm(tm& time)` change and that will be replaced by your changes.

basilgello replied:
OK, thanks for checks! Then let's make this PR in. I thought the overlap will be huge in TestDateTime :)

neo1973 committed:
TestDateTime: Extend struct tm tests to catch DST problems
